### ISSUE TITLE
drop support for python 3.8

### ISSUE THREAD

**Issue by baywet** [microsoft/kiota-python#349]
Python 3.8 has [fallen out of support](https://endoflife.date/python) 2 days ago, which means it's not getting anymore security updates.
This also has consequences for dependency management since an increasing number of dependencies have moved away from it, and we're not able to get the new versions anymore.
Consider making this change at the same time as #341, and after #302 has been implemented.

TODO:
- [x] remove 3.8 from the toml file
- [x] remove 3.8 from the matrix of the workflows
- [x] remove 3.8 from the policy definition if #327 has been implemented.

IMPORTANT: this is a source breaking change for 3.8 users (they'll need to upgrade to a newer version of python), make sure to include a bang in your message commit (e.g. `feat!:`) so a major version bump is made.

We do not need to major bump kiota itself since the generator will emit for the same API surface and we're within our support policy.

After the PR for this is merged but BEFORE the changes are released we also need to go back to any closed (not merged) dependabot PR and ask the bot to recreate the pull requests (`@dependabot recreate`) so we upgrade to the latest version of the dependencies.

**Comment by baywet** [comment]
additionally this documentation needs to be updated https://learn.microsoft.com/en-us/openapi/kiota/quickstarts/python#required-tools

**Comment by baywet** [comment]
And this one https://learn.microsoft.com/en-us/openapi/kiota/tutorials/python-azure?tabs=portal

**Comment by baywet** [comment]
to add data points for this one: it seems that libraries are perfectly fine with bumping the required minor version of python in a patch/minor release.
Here are some examples:
- https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/723
- https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/730
- https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/729
- https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/722
- https://github.com/microsoftgraph/msgraph-sdk-python-core/pull/720

Which means this is most likely was the eco-system expects, and it should be fine to follow that practice on our end.

**Comment by andrueastman** [comment]
Seeing the same in changelogs such as 

- https://github.com/agronholm/anyio/releases/tag/4.6.0
- https://setuptools.pypa.io/en/stable/history.html#v75-4-0
- https://github.com/pylint-dev/pylint/releases/tag/v3.3.0

**Comment by baywet** [comment]
Looking at the [Azure SDK guidelines](https://azure.github.io/azure-sdk/python_design.html#supported-python-versions), they are just providing a (outdated?) target version. No upgrade guidance. I have asked a question in the internal channels to see if any Python architect tells us following what the ecosystem is doing here is a bad idea for some reason.

**Comment by Ndiritu** [comment]
We could also consider dropping the[ deprecated type hint](https://docs.python.org/3/library/typing.html#deprecated-aliases) aliases in favor of the recommended types for Python 3.9+

**Comment by andrueastman** [comment]
> I think that with all the evidence we've collected, it'd be ok to remove support for 3.8 in a minor bump. What do you think @andrueastman @Ndiritu ? If we're in agreement, we can probably proceed with the work.

Agree. We can drop the version with a minor bump given the evidence collected.

**Comment by baywet** [comment]
We should now expedite this, dependabot (and most likely the associated security features) is dropping 3.8 support in less than a month
https://github.blog/changelog/2025-01-06-closing-down-notice-dependabot-will-no-longer-support-python-version-3-8

**Comment by andrueastman** [comment]
Taking a look into this.

**Comment by andrueastman** [comment]
I believe https://github.com/microsoft/kiota-python/pull/433 should be okay to review now. 

On investigating, it looks like we are also emitting the deprecated type hints in the generated code, so I've also authored a generator PR at https://github.com/microsoft/kiota/pull/5974.

My suggestion for review/release order here would be to 

- [ ] - Review and merge in https://github.com/microsoft/kiota-python/pull/433
- [ ] - Validate needed dependabot updates then release next minor (tentatively v1.7.0)
- [ ] - Review and merge Generation PR https://github.com/microsoft/kiota/pull/5974 with updated dependency of 1.7.0 
- [ ] - Review and merge Updated validation client at https://github.com/microsoft/kiota-python/pull/434 (updated with any feedback from above)
- [ ] - Review and merge docs PR at https://github.com/MicrosoftDocs/openapi-docs/pull/141
